src/datastore/Datastore.cc

```diff
@@ -1113,59 +1111,115 @@ bool Datastore::get_avail_mb(long long &avail) const
 
-    rc = Nebula::instance().get_ds_conf_attribute(ds_mad, vatt);
+    int rc = Nebula::instance().get_ds_conf_attribute(mad, vatt);
 
     if ( rc != 0 )
     {
         // No DS_MAD_CONF is available for this DS_MAD.
-        // Assuming this DS is not PERSISTENT_ONLY
-        return false;
+        return defval;
     }
 
-    vatt->vector_value("PERSISTENT_ONLY", persistent_only);
+    T value;
 
-    return persistent_only;
-};
+    rc = vatt->vector_value(name, value);
 
-/* ------------------------------------------------------------------------ */
-/* ------------------------------------------------------------------------ */
+    if ( rc != 0 )
+    {
+        // Attribute missing in DS_MAD_CONF
+        return defval;
+    }
+
+    return value;
+}
+
+/* -------------------------------------------------------------------------- */
+
+bool Datastore::is_persistent_only() const
+{
+    return ds_conf_value(ds_mad, "PERSISTENT_ONLY", false);
+}
+
+bool Datastore::is_concurrent_forget() const
+{
+    return ds_conf_value(ds_mad, "CONCURRENT_FORGET", false);
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
 
 int Datastore::get_tm_mad_targets(const string &tm_mad, string& ln_target,
         string& clone_target, string& disk_type) const
 {
-    if (!tm_mad.empty())
+    if (tm_mad.empty())
     {
-        string tm_mad_t = one_util::trim(tm_mad);
-        one_util::toupper(tm_mad_t);
+        return 0;
+    }
 
-        get_template_attribute("CLONE_TARGET_" + tm_mad_t, clone_target);
+    string tm_mad_t = one_util::trim(tm_mad);
+    one_util::toupper(tm_mad_t);
 
-        if (clone_target.empty())
-        {
-            return -1;
-        }
+    get_template_attribute("CLONE_TARGET_" + tm_mad_t, clone_target);
+
+    if (clone_target.empty())
+    {
+        return -1;
+    }
 
-        get_template_attribute("LN_TARGET_" + tm_mad_t, ln_target);
+    get_template_attribute("LN_TARGET_" + tm_mad_t, ln_target);
 
-        if (ln_target.empty())
-        {
-            return -1;
-        }
+    if (ln_target.empty())
+    {
+        return -1;
+    }
 
-        get_template_attribute("DISK_TYPE_" + tm_mad_t, disk_type);
+    get_template_attribute("DISK_TYPE_" + tm_mad_t, disk_type);
 
-        if (disk_type.empty())
-        {
-            return -1;
-        }
+    if (disk_type.empty())
+    {
+        return -1;
     }
 
     return 0;
 }
+
+/* ------------------------------------------------------------------------ */
+/* ------------------------------------------------------------------------ */
+
+Image::DiskType Datastore::context_disk_type() const
+{
+    Image::DiskType ctxt_dt = Image::FILE;
+    string          ctxt_dt_s;
+
+    get_template_attribute("CONTEXT_DISK_TYPE", ctxt_dt_s);
+
+    if (!ctxt_dt_s.empty())
+    {
+        ctxt_dt = Image::str_to_disk_type(ctxt_dt_s);
+    }
+
+    switch(ctxt_dt)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The switch statement here is missing a default statement. The default statement will catch any case values that are not explicitly handled by a case statement.  This prevents bugs if a case value is accidentally forgotten or new enum values are added later